presto-main/src/test/java/com/facebook/presto/operator/TestDictionaryLookupJoinPageBuilder.java

```diff
@@ -0,0 +1,227 @@
+        JoinProbe probe = simpleJoinProbeFactory.createJoinProbe(page);
+        LookupSource lookupSource = new TestLookupSource(ImmutableList.of(BIGINT, BIGINT), page);
+        LookupJoinPageBuilder lookupJoinPageBuilder = new DictionaryLookupJoinPageBuilder(ImmutableList.of(BIGINT, BIGINT));
+
+        int joinPosition = 0;
+        while (!lookupJoinPageBuilder.isFull() && probe.advanceNextPosition()) {
+            lookupJoinPageBuilder.appendRow(probe, lookupSource, joinPosition++);
+            lookupJoinPageBuilder.appendNullForBuild(probe);
+        }
+        assertFalse(lookupJoinPageBuilder.isEmpty());
+
+        Page output = lookupJoinPageBuilder.build(probe);
+        assertEquals(output.getChannelCount(), 4);
+        for (int i = 0; i < output.getPositionCount(); i++) {
+            assertFalse(output.getBlock(0).isNull(i));
+            assertFalse(output.getBlock(1).isNull(i));
+            assertEquals(output.getBlock(0).getLong(i, 0), i / 2);
+            assertEquals(output.getBlock(1).getLong(i, 0), i / 2);
+            if (i % 2 == 0) {
+                assertFalse(output.getBlock(2).isNull(i));
+                assertFalse(output.getBlock(3).isNull(i));
+                assertEquals(output.getBlock(2).getLong(i, 0), i / 2);
+                assertEquals(output.getBlock(3).getLong(i, 0), i / 2);
+            }
+            else {
+                assertTrue(output.getBlock(2).isNull(i));
+                assertTrue(output.getBlock(3).isNull(i));
+            }
+        }
+        assertTrue(lookupJoinPageBuilder.toString().contains("positionCount=" + output.getPositionCount()));
+
+        lookupJoinPageBuilder.reset();
+        assertTrue(lookupJoinPageBuilder.isEmpty());
+    }
+
+    @Test
+    public void testDifferentPositions()
+            throws Exception
+    {
+        int entries = 100;
+        BlockBuilder blockBuilder = BIGINT.createFixedSizeBlockBuilder(entries);
+        for (int i = 0; i < entries; i++) {
+            BIGINT.writeLong(blockBuilder, i);
+        }
+        Block block = blockBuilder.build();
+        Page page = new Page(block);
+        SimpleJoinProbeFactory simpleJoinProbeFactory = new SimpleJoinProbeFactory(ImmutableList.of(BIGINT), new int[]{0}, ImmutableList.of(0), OptionalInt.empty());
+        LookupSource lookupSource = new TestLookupSource(ImmutableList.of(BIGINT), page);
+        LookupJoinPageBuilder lookupJoinPageBuilder = new DictionaryLookupJoinPageBuilder(ImmutableList.of(BIGINT));
+
+        // empty
+        JoinProbe probe = simpleJoinProbeFactory.createJoinProbe(page);
+        Page output = lookupJoinPageBuilder.build(probe);
+        assertEquals(output.getChannelCount(), 2);
+        assertEquals(output.getPositionCount(), 0);
+        lookupJoinPageBuilder.reset();
+
+        // the probe covers non-sequential positions
```

Comment:
Can you add assertions for block types?  Specifically, we should be getting a dictionary block in some cases an not in others.